File: 06-SVM Algorithm/ex6.py
from __future__ import division
import numpy as np
from scipy import io
import matplotlib.pyplot as plt
from sklearn import svm

def plot_data(X, y):
    pos = np.nonzero(y == 1)
    neg = np.nonzero(y == 0)
    plt.plot(X[pos, 0], X[pos, 1], linestyle='', marker='+', color='k')
    plt.plot(X[neg, 0], X[neg, 1], linestyle='', marker='o', color='y')

def visualize_boundary_linear(X, y, clf):
    plot_data(X, y)
    coef = clf.coef_.ravel()
    intercept = clf.intercept_.ravel()

    xp = np.linspace(np.min(X[:, 0]), np.max(X[:, 0]), 100)
    yp = -1.0 * (coef[0] * xp + intercept[0]) / coef[1]
    plt.plot(xp, yp, linestyle='-', color='b')

def visualize_boundary(X, y, clf):
    plot_data(X, y)
    x1_plot = np.linspace(np.min(X[:, 0]), np.max(X[:, 0]), 100)
    x2_plot = np.linspace(np.min(X[:, 1]), np.max(X[:, 1]), 100)
    X1, X2 = np.meshgrid(x1_plot, x2_plot)
    vals = np.zeros(X1.shape)

    for i in range(X1.shape[1]):
        X_tmp = np.hstack((X1[:, i:i + 1], X2[:, i:i + 1]))
        vals[:, i] = clf.predict(X_tmp)
    plt.contour(X1, X2, vals, levels=[0])

def dataset1():
    data = io.loadmat('ex6data1.mat')
    X = data['X']
    y = data['y'].ravel()

    clf = svm.LinearSVC(C=1)
    clf.fit(X, y)
    print('Accuracy score:', clf.score(X, y))
    plt.figure(1)
    visualize_boundary_linear(X, y, clf)
    plt.show()

def dataset2():
    data = io.loadmat('ex6data2.mat')
    X = data['X']
    y = data['y'].ravel()

    clf = svm.SVC(C=100, kernel='rbf', gamma=10)
    clf.fit(X, y)
    print('Accuracy score:', clf.score(X, y))
    plt.figure(2)
    visualize_boundary(X, y, clf)
    plt.show()

def data3_params(X, y, X_val, y_val):
    C_cands = [0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30]
    gamma_cands = [0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30]
    max_score = -1
    C_pick = -1
    gamma_pick = -1
    for C in C_cands:
        for gamma in gamma_cands:
            clf = svm.SVC(C=C, kernel='rbf', gamma=gamma)
            clf.fit(X, y)
            score = clf.score(X_val, y_val)
            if score > max_score:
                max_score = score
                C_pick = C
                gamma_pick = gamma

    return C_pick, gamma_pick

def dataset3():
    data = io.loadmat('ex6data3.mat')
    X = data['X']
    y = data['y'].ravel()
    X_val = data['Xval']
    y_val = data['yval'].ravel()

    C_pick, gamma_pick = data3_params(X, y, X_val, y_val)

    clf = svm.SVC(C=C_pick, kernel='rbf', gamma=gamma_pick)
    clf.fit(X, y)
    print('Accuracy score:', clf.score(X, y))
    plt.figure(3)
    visualize_boundary(X, y, clf)
    plt.show()

if __name__ == "__main__":

    dataset1()
    dataset2()
    dataset3()

File: 06-SVM Algorithm/ex6.m
load('ex6data1.mat');

function plotData(X, y)
  pos = find(y == 1); neg = find(y == 0);
  plot(X(pos, 1), X(pos, 2), 'k+','LineWidth', 1, 'MarkerSize', 7)
  hold on;
  plot(X(neg, 1), X(neg, 2), 'ko', 'MarkerFaceColor', 'y', 'MarkerSize', 7)
  hold off;
endfunction

function sim = linearKernel(x1, x2)
  x1 = x1(:); x2 = x2(:);
  sim = x1' * x2;
endfunction

function visualizeBoundaryLinear(X, y, model)
  w = model.w;
  b = model.b;
  xp = linspace(min(X(:,1)), max(X(:,1)), 100);
  yp = - (w(1)*xp + b)/w(2);
  plotData(X, y);
  hold on;
  plot(xp, yp, '-b'); 
  hold off
endfunction

plotData(X, y);
C = 1;
model = svmTrain(X, y, C, @linearKernel, 1e-3, 20);
visualizeBoundaryLinear(X, y, model);

function sim = gaussianKernel(x1, x2, sigma)
  x1 = x1(:); x2 = x2(:);
  sim = 0;
  sim = exp(-sum((x1-x2).^2)/(2*(sigma ^ 2)));
endfunction

fprintf('\nEvaluating the Gaussian Kernel ...\n')
x1 = [1 2 1]; x2 = [0 4 -1]; sigma = 2;
sim = gaussianKernel(x1, x2, sigma);

fprintf(['Gaussian Kernel between x1 = [1; 2; 1], x2 = [0; 4; -1], sigma = 2 :' ...
         '\n\t%f\n(this value should be about 0.324652)\n'], sim);

function visualizeBoundary(X, y, model, varargin)
  plotData(X, y)
  x1plot = linspace(min(X(:,1)), max(X(:,1)), 100)';
  x2plot = linspace(min(X(:,2)), max(X(:,2)), 100)';
  [X1, X2] = meshgrid(x1plot, x2plot);
  vals = zeros(size(X1));
  for i = 1:size(X1, 2)
    this_X = [X1(:, i), X2(:, i)];
    vals(:, i) = svmPredict(model, this_X);
  endfor
  hold on
  contour(X1, X2, vals, [0 0], 1 , 'linecolor', 'blue');
  hold off;
endfunction
         
load('ex6data2.mat');
plotData(X, y);
C = 1;
sigma = 0.1;
model= svmTrain(X, y, C, @(x1, x2) gaussianKernel(x1, x2, sigma)); 
visualizeBoundary(X, y, model);

function [C, sigma] = dataset3Params(X, y, Xval, yval)  
  possibles = [0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30];
  error = zeros(size(possibles, 2),size(possibles, 2));
  for i=1:size(possibles, 2)
    for j=1:size(possibles, 2)
      C_pos = possibles(i);
      S_pos = possibles(j);
      model = svmTrain(X, y, C_pos, @(x1, x2) gaussianKernel(x1, x2, S_pos));
      y_pred = svmPredict(model, Xval);
      error(i,j) = mean(double(y_pred ~= yval));
    endfor
  endfor
  [r,c] = find(error == min(min(error)));
  C = possibles(r);
  sigma = possibles(c);
endfunction

load('ex6data3.mat');
plotData(X, y);
[C, sigma] = dataset3Params(X, y, Xval, yval);
model= svmTrain(X, y, C, @(x1, x2) gaussianKernel(x1, x2, sigma));
visualizeBoundary(X, y, model);